### PR TITLE
[ORCA-135] Enable remote editing and testing of this repo on GitHub Codespaces

### PR DESCRIPTION
**Pull request by BrunoGrandePhD**
Here's a first draft of the `devcontainer.json` for running Docker-in-Docker on Ubuntu. This could probably me made more efficient, but we can start here since it works! I also updated the README to reflect the updated usage. 

**Edit:** I intentionally removed the `requirements.txt` file in favor of `Pipfile`, which allows us to separately specify the "target dependencies" (_e.g._ `orca`) from their dependencies (which are listed in `Pipfile.lock`).

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+	"portsAttributes": {
+		"8080": {
+			"label": "airflow-ui",
+			"onAutoForward": "silent"
+		}
+	},
+
+	// Use 'postCreateCommand' to run commands after the container is created.
+	"postCreateCommand": "git config --global --add safe.directory /workspaces/orca-recipes-airflow",
```
Comment (BrunoGrandePhD) [inline .devcontainer/devcontainer.json]: @tschaffter: I was getting the following error when I tried `git push`. They suggested that I ran this command, which indeed fixed the issue, but I'm hoping to address the root cause. 

I noticed that the folders in the repo are owned by `root` whereas my user is `vscode`. Did you encounter this and if so, how did you address it? 

```
Git: fatal: detected dubious ownership in repository at '/workspaces/orca-recipes-airflow'
```

Comment (tschaffter) [reply .devcontainer/devcontainer.json]: I [create and configure the user `vscode`](https://github.com/Sage-Bionetworks/sage-monorepo/blob/869abca18aff11af9fa5d682b4ea1a9b494dfcf6/tools/devcontainers/sage/.devcontainer/Dockerfile#L74-L78) in the Dockerfile of the dev container.
I don't remember encountering the issue above.

**Review by thomasyu888 — COMMENTED** [review]
@BWMac also brings up a good point about how we are going to deal with AWS secrets connections using this method.

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/ubuntu
+{
+	"name": "Ubuntu",
```
Comment (thomasyu888) [inline .devcontainer/devcontainer.json]: Running in an issue when running locally:

```
#0 18.47 ERROR: Could not install packages due to an OSError: [Errno 28] No space left on device: '/tmp/pip-unpack-jcb8wvdf'
```

Just had to remove existing images

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# Contribution Guidelines
+
+## Infrastructure
+
+We have two Airflow servers:
+
+<!-- Do we need `airflow-dev` if we have DevContainers? -->
+1. `airflow-dev`: Hosted in the `dnt-dev` AWS account. Deployed manually. Has a public IP address (accessible using the VPN).
```
Comment (thomasyu888) [inline CONTRIBUTING.md:8]: Eventually, can this be a non-manual deployment using CI/CD? OR is there value in this (as mentioned on slack)?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	* Deployed from the `develop` branch in this repository.
+2. `airflow-prod`: Hosted in the `dpe-prod` AWS account. Deployed using CloudFormation. Only has a private IP address (accessible via SSM port forwarding).
+	* Deployed from the `main` branch in this repository.
+
+There is a helper script in this repository for accessing these Airflow servers.
+
+## Development
+
+To develop on this repository, it's recommended that you use the Dev Containers setup online using GitHub Codespaces. If you must, you can use the VS Code "Dev Containers" extension to develop locally, but this would require you to configure your secrets (_e.g._ using the `.env` file).
```
Comment (thomasyu888) [inline CONTRIBUTING.md:17]: Does github codespaces allow you to save secrets within the repo?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+## Secrets
+
+Airflow secrets (_e.g._ connections and variables) are stored in the following locations:
+
+<!-- Maybe we can store all secrets (both test and prod) in `dpe-prod`? -->
+- `dpe-prod` AWS account: Production-level secrets (_e.g._ actual Tower workspaces)
```
Comment (thomasyu888) [inline CONTRIBUTING.md:24]: What is the process for getting a secret stored?

**Comment by BWMac** [comment]
Just wanted to drop a comment that I was able to run this dev container on my machine, so looks like there are no M1 mac compatibility issues

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (tschaffter) [reply .devcontainer/devcontainer.json]: @BrunoGrandePhD Did you address the root cause of this issue?

File: CONTRIBUTING.md (hunk 3 above)
Comment (BrunoGrandePhD) [reply CONTRIBUTING.md:8]: A follow-up PR will automate the deployment.

File: CONTRIBUTING.md (hunk 4 above)
Comment (BrunoGrandePhD) [reply CONTRIBUTING.md:17]: Yes, but I still need to test that out.

File: CONTRIBUTING.md (hunk 5 above)
Comment (BrunoGrandePhD) [reply CONTRIBUTING.md:24]: Once I figure that out, I'll document it here.

**Review by tschaffter — APPROVED** [review]

File: README.md [hunk 6]
```diff
@@ -1,22 +1,41 @@
...
+This assumes that you have Docker installed with [Docker Compose V2](https://docs.docker.com/compose/compose-v2/). It's recommended that you leverage the included Dev Container definition (_i.e._ `devcontainer.json`) to standardize your development environment. You can use the [Dev Containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) VS Code extension or GitHub Codespaces.
+
+```console
+# Duplicate example `.env` file and edit as needed
+cp .env.example .env
+docker compose up --build --detach
 ```
 
-### Start Docker Containers
+After you're done editing your DAG, you can restart the containers so Airflow can pick up on the latest version as follows.
```
Comment (BWMac) [inline README.md]: Technically changes in the `dags` directory will be picked up automatically so a refresh on the UI web page will do the trick

File: README.md [hunk 7]
```diff
@@ -1,22 +1,41 @@
...
+
+```console
+docker compose restart
 ```
-docker-compose up
+
+If you edit `Dockerfile`, `docker-compose.yaml`, or `Pipfile`/`Pipfile.lock`, you'll need to rebuild the containers as follows.
```
Comment (BWMac) [inline README.md]: editing `airflow.cfg` or any sort of `.env` file will require a rebuild as well

**Review by BWMac — APPROVED** [review]

File: README.md (hunk 6 above)
Comment (BrunoGrandePhD) [reply README.md]: Good to know!

File: README.md (hunk 7 above)
Comment (BrunoGrandePhD) [reply README.md]: Good catch!

**Pull request merged**